app/controllers/incomeSources/add/IncomeSourceAddedController.scala

```diff
@@ -0,0 +1,142 @@
+                                            val obligationsView: IncomeSourceAddedObligations,
+                                            nextUpdatesService: NextUpdatesService)
+                                           (implicit val appConfig: FrontendAppConfig,
+                                            implicit val itvcErrorHandlerAgent: AgentItvcErrorHandler,
+                                            implicit override val mcc: MessagesControllerComponents,
+                                            val ec: ExecutionContext,
+                                            dateService: DateServiceInterface)
+  extends ClientConfirmedController with I18nSupport with FeatureSwitching with IncomeSourcesUtils {
+
+  def show(incomeSourceId: String, incomeSourceType: IncomeSourceType): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(isAgent = false, incomeSourceId, incomeSourceType)
+  }
+
+  def showAgent(incomeSourceId: String, incomeSourceType: IncomeSourceType): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+          implicit mtdItUser =>
+            handleRequest(isAgent = true, incomeSourceId, incomeSourceType)
+        }
+  }
+
+  private def getBackUrl(incomeSourceId: String, isAgent: Boolean, incomeSourceType: IncomeSourceType): String = {
+    val baseRoute = incomeSourceType match {
+      case SelfEmployment =>
+        if (isAgent) controllers.incomeSources.add.routes.BusinessReportingMethodController.showAgent _ else
+          controllers.incomeSources.add.routes.BusinessReportingMethodController.show _
+      case UkProperty =>
+        if (isAgent) controllers.incomeSources.add.routes.UKPropertyReportingMethodController.showAgent _ else
+          controllers.incomeSources.add.routes.UKPropertyReportingMethodController.show _
+      case ForeignProperty =>
+        if (isAgent) controllers.incomeSources.add.routes.ForeignPropertyReportingMethodController.showAgent _ else
+          controllers.incomeSources.add.routes.ForeignPropertyReportingMethodController.show _
+    }
+    baseRoute(incomeSourceId).url
+  }
+
+  private def handleRequest(isAgent: Boolean, incomeSourceId: String, incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
+    withIncomeSourcesFS {
+      val backUrl = getBackUrl(incomeSourceId, isAgent, incomeSourceType)
+       incomeSourceDetailsService.getIncomeSourceFromUser(incomeSourceType, incomeSourceId) match {
+        case Some((startDate, businessName)) =>
+          val showPreviousTaxYears: Boolean = startDate.isBefore(dateService.getCurrentTaxYearStart())
+          incomeSourceType match {
```

Comment:
if you can move this part (under incomeSourceType) - under separate private method for better code readability